### PR TITLE
Update DoSetMachineDeployment to fix np name issue

### LINKED ISSUE #1776: Unsafe regular expressions in machine_deployment.go cause wrong node-pools to be deleted and scaled
**Bug description**

Unsafe regular expressions in machine_deployment.go cause wrong node-pools to be deleted and scaled. Extremely dangerous and renders the product unusable for us.

**Affected product area (please put an X in all that apply)**

- [X] APIs
- [ ] Addons
- [X] CLI
- [ ] Docs
- [ ] IAM
- [ ] Installation
- [ ] Plugin
- [X] Security
- [ ] Test and Release
- [X] User Experience

**Expected behavior**

tanzu cluster scale and tanzu cluster node-pool commands should act upon the node-pool specified by the administrator, not some other node-pool that has a name that satisfied the flawed regex in the code.

**Steps to reproduce the bug**

*Prep*

Create a node-pool called gw-md-0.

*Scaling the wrong pool*

```
tanzu cluster node-pool list foo-wl-01
  NAME     NAMESPACE  PHASE    REPLICAS  READY  UPDATED  UNAVAILABLE
  gw-md-0  default    Running  1         1      1        0
  md-0     default    Running  1         1      1        0
  md-1     default    Running  1         1      1        0
  md-2     default    Running  1         1      1        0

tanzu cluster scale foo-wl-01 -p md-0 -w 3
Workload cluster 'foo-wl-01' is being scaled

tanzu cluster node-pool list foo-wl-01
  NAME     NAMESPACE  PHASE      REPLICAS  READY  UPDATED  UNAVAILABLE
  gw-md-0  default    ScalingUp  3         1      3        2
  md-0     default    Running    1         1      1        0
  md-1     default    Running    1         1      1        0
  md-2     default    Running    1         1      1        0
```

*Deleting the wrong pool*

```
tanzu cluster node-pool list foo-wl-01
  NAME     NAMESPACE  PHASE    REPLICAS  READY  UPDATED  UNAVAILABLE
  gw-md-0  default    Running  1         1      1        0
  md-0     default    Running  1         1      1        0
  md-1     default    Running  1         1      1        0
  md-2     default    Running  1         1      1        0

tanzu cluster node-pool delete foo-wl-01 -n md-0

tanzu cluster node-pool list foo-wl-01
  NAME  NAMESPACE  PHASE    REPLICAS  READY  UPDATED  UNAVAILABLE
  md-0  default    Running  1         1      1        0
  md-1  default    Running  1         1      1        0
  md-2  default    Running  1         1      1        0
```

**Version** (include the SHA if the version is not obvious)

v0.11.1 and still in main

**Environment where the bug was observed (cloud, OS, etc)**

VSphere

**Relevant Debug Output (Logs, manifests, etc)**

These problems occur due to these loose regexes:

https://github.com/vmware-tanzu/tanzu-framework/blob/v0.11.1/pkg/v1/tkg/client/machine_deployment.go#L125
https://github.com/vmware-tanzu/tanzu-framework/blob/v0.11.1/pkg/v1/tkg/client/machine_deployment.go#L325

### PR DESCRIPTION
**Pull request by tenczar**
Fixes an issue where DoSetMachineDeployment will select the incorrect
machinedeployment based on its name.

### Which issue(s) this PR fixes

Fixes #1776 

### Release note

```release-note
Fixes an issue where the incorrect node pool will be updated when the name is a substring of another node pool.
```

### PR Checklist

- [ ] Squash the commits into one or a small number of logical commits
      
- [ ] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [ ] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

**Review by prkalle — APPROVED** [review]
LGTM.

**Comment by nickperry** [comment]
@tenczar thank you

File: pkg/v1/tkg/client/machine_deployment_test.go [hunk 1]
```diff
@@ -1181,6 +1181,25 @@ var _ = Describe("Machine Deployment", func() {
 						},
 					},
 				}
+				md2 = capi.MachineDeployment{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "test-cluster-np-1",
```
Comment (vuil) [inline pkg/v1/tkg/client/machine_deployment_test.go:1186]: not sure if it is intentional, but it is weird to set both to the same name, also weird later to modify md2's name to match np1Name

**Review by vuil — COMMENTED** [review]
change looks good, but the test is a little confusing

**Review by vuil — APPROVED** [review]
won't block the change due to test, but if we do merge this, please update the test case for better clarity.

**Pull request merged**